central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,646 @@
+		Schema:                  schema,
+	}
+	s.numRoutines.Store(0)
+	return s
+}
+
+func (s *scheduler) UpsertReportSchedule(reportConfig *storage.ReportConfiguration) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	cronSpec := ""
+	var err error
+	var entryID cron.EntryID
+	if reportConfig.GetSchedule() != nil {
+		cronSpec, err = schedule.ConvertToCronTab(reportConfig.GetSchedule())
+		if err != nil {
+			return err
+		}
+		entryID, err = s.cron.AddFunc(cronSpec, s.reportClosure(reportConfig))
+		if err != nil {
+			return err
+		}
+	}
+
+	// Remove the old entry if this is an update
+	if oldEntryID, ok := s.reportConfigToEntryIDs[reportConfig.GetId()]; ok {
+		s.cron.Remove(oldEntryID)
+	}
+	// Only add to entries if this is still a scheduled report config
+	if cronSpec != "" {
+		s.reportConfigToEntryIDs[reportConfig.GetId()] = entryID
+	}
+	return nil
+}
+
+func (s *scheduler) reportClosure(reportConfig *storage.ReportConfiguration) func() {
+	return func() {
+		log.Infof("Submitting scheduled report request for '%s' at %v", reportConfig.GetName(), time.Now().Format(time.RFC850))
+		_, err := s.SubmitReport(&ReportRequest{
+			ReportConfig: reportConfig,
+			ReportMetadata: &storage.ReportMetadata{
+				ReportConfigId: reportConfig.Id,
+				ReportStatus: &storage.ReportStatus{
+					RunState:                 storage.ReportStatus_WAITING,
+					ReportRequestType:        storage.ReportStatus_SCHEDULED,
+					ReportNotificationMethod: storage.ReportStatus_EMAIL,
+				},
+			},
+		}, false)
+		if err != nil {
+			log.Errorf("Error submitting scheduled report request for '%s': %s", reportConfig.GetName(), err)
+		}
+	}
+}
+
+func (s *scheduler) RemoveReportSchedule(reportConfigID string) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	oldEntryID, exists := s.reportConfigToEntryIDs[reportConfigID]
+	if exists {
+		s.cron.Remove(oldEntryID)
+		delete(s.reportConfigToEntryIDs, reportConfigID)
+	}
+}
+
+func (s *scheduler) SubmitReport(request *ReportRequest, reSubmission bool) (string, error) {
+	request.Ctx = selectContext(request)
+	var err error
+	if request.ReportMetadata.GetReportStatus().GetReportRequestType() == storage.ReportStatus_ON_DEMAND {
+		userHasAnotherReport, err := s.doesUserHavePendingReport(request.Ctx, request.ReportConfig.GetId(), request.ReportMetadata.GetRequester().GetId())
+		if err != nil {
+			return "", err
+		}
+		if userHasAnotherReport {
+			return "", errors.Errorf("User already has a running report for config ID %s", request.ReportConfig.GetId())
+		}
+	}
+
+	request.ReportMetadata.ReportStatus.RunState = storage.ReportStatus_WAITING
+	request.ReportMetadata.ReportStatus.QueuedAt = types.TimestampNow()
+	if !reSubmission {
+		request.ReportMetadata.ReportId, err = s.reportMetadataStore.AddReportMetadata(request.Ctx, request.ReportMetadata)
+	} else {
+		err = s.reportMetadataStore.UpdateReportMetadata(request.Ctx, request.ReportMetadata)
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	s.lock.Lock()
```

Comment:
This lock is pulling double duty. It's locking on access to `cron.Cron` and `reportConfigToEntryIDs` already and now it's locking `queuedReports`. Do you want access to `queuedReports` blocked because another thread is updating `reportConfigToEntryIDs`?